### ISSUE TITLE
Multiple definitions of `cxxopts::values::integer_pattern`

### ISSUE THREAD

**Issue by SylvainCorlay** [jarro2783/cxxopts#69]
Creating a global value does makes one version of it per compilation unit, which causes problems at link time when cxxopts is used in more than one compilation units.

Instead of doing

```cpp
namespace
{
  std::basic_regex<char> integer_pattern
    ("(-)?(0x)?([1-9a-zA-Z][0-9a-zA-Z]*)|(0)");
}
```

you probably should have

```cpp
inline const std::basic_regex<char>& integer_pattern()
{
    static const std::basic_regex<char> pattern = "(-)?(0x)?([1-9a-zA-Z][0-9a-zA-Z]*)|(0)";
    return pattern;
}
```

i.e. make a singleton.

**Comment by SylvainCorlay** [comment]
cc @JohanMabille

**Comment by jarro2783** [comment]
I thought putting it in an anonymous namespace would fix it, and I also thought that I tested this.

Can you show me a snippet of code that breaks?

**Comment by SylvainCorlay** [comment]
OK, I can put one together tomorrow. This came up in a rather large project.

Basically, we are working on a C++ kernel for the Jupyter notebook, backed by the cling interpreter:

![peek 2017-08-16 12-45](https://user-images.githubusercontent.com/2397974/29896722-1fedb4b6-8dac-11e7-96a2-7aaec8c56fe3.gif)

We have a notion of "cell magic" that allows user to enter non-c++ commands to perform some operations (fetching help on a certain module etc).

We use cxxopts to parse options for these commands. So basically, our usecase is quite difference from the usual argument parsing for a command-line application but cxxopts does the job, hence the multiple compilation units.

**Comment by jarro2783** [comment]
This really sounds like a bug in the compiler if anything. I'm pretty sure that according to the standard, something declared in an anonymous namespace has internal linkage in the translation unit, so it shouldn't result in multiple references.

If you can give me some code that reproduces the problem then I can have a look at it.

**Comment by SylvainCorlay** [comment]
The code that breaks is the compilation of [xeus-cling](https://github.com/QuantStack/xeus-cling).

Building it locally is a bit cumbersome because it has a lot of dependencies which need to be built or for which you will need to get a binary package. (For this reason, I use the conda package manager for which I have built all the dependency tree).

I do get multiple versions of gcc (4.9, 5.x anx 6.x) to complain `cxxopts::values::integer_pattern`. However, I don't get why this should not work.

FYI, the version with a singleton does work.

**Comment by SylvainCorlay** [comment]
and also prevents recompiling the regex.

**Comment by jarro2783** [comment]
In fact I'm not sure why you get an error with `integer_pattern` and not `option_matcher`, `option_specifier`.

**Comment by jarro2783** [comment]
I'm trying to build `xeus-cling`, to reproduce this. I followed the instructions on that repo to install `conda`, and cloned the `xeus-cling` source, but I'm stuck here:
```
cmake -DCMAKE_INSTALL_PATH=$HOME/miniconda3 -DCMAKE_INSTALL_LIBDIR=$HOME/miniconda3 .
-- The C compiler identification is GNU 7.1.1
-- The CXX compiler identification is GNU 7.1.1
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found LLVM_CONFIG as /home/jbeck/miniconda3/bin/llvm-config
CMake Error at /usr/lib/cmake/clang/ClangTargets.cmake:419 (message):
  The imported target "clangApplyReplacements" references the file

     "/usr/lib/libclangApplyReplacements.a"
```

What am I missing here?

**Comment by jarro2783** [comment]
I was never able to reproduce this, but it will be fixed when I address #281
